backends/graphics/atari/atari-graphics.cpp

```diff
@@ -0,0 +1,998 @@
+		copyRectToSurfaceWithKey(_cursor.surface, _cursor.dstRect.left, _cursor.dstRect.top,
+			_screenSurface, _cursor.srcRect, _cursor.keycolor);
+
+		_cursor.positionChanged = _cursor.surfaceChanged = false;
+		_oldCursorRect = _cursor.dstRect;
+
+		updated = true;
+	}
+
+	return updated;
+}
+
+bool AtariGraphicsManager::updateDoubleAndTripleBuffer()
+{
+	bool updated = false;
+	bool drawCursor = _cursor.isModified();
+
+	if (_screenModified) {
+		drawCursor = true;
+
+		copySurfaceToSurface(_chunkySurface, _screenSurface);
+
+		// updated in screen swapping
+		//_screenModified = false;
+		updated = true;
+	}
+
+	if (_cursor.outOfScreen)
+		return updated;
+
+	// render directly to the screen to be swapped (so we don't have to refresh full screen when only cursor moves)
+	Graphics::Surface frontBufferScreenSurface;
+	frontBufferScreenSurface.init(_screenSurface.w, _screenSurface.h, _screenSurface.pitch,
+		_screenAligned[_screenModified ? BACK_BUFFER1 : FRONT_BUFFER], _screenSurface.format);
+
+	if ((_cursor.positionChanged || !_cursor.visible) && !_oldCursorRect.isEmpty() && !_screenModified) {
+		alignRect(_chunkySurface, _oldCursorRect);
+		copyRectToSurface(_chunkySurface, _oldCursorRect.left, _oldCursorRect.top,
+			frontBufferScreenSurface, _oldCursorRect);
+
+		_oldCursorRect = Common::Rect();
+
+		updated = true;
+	}
+
+	if (drawCursor && _cursor.visible) {
+		//debug("Redraw cursor (double/triple): %d %d %d %d", _cursor.dstRect.left, _cursor.dstRect.top, _cursor.dstRect.width(), _cursor.dstRect.height());
+
+		copyRectToSurfaceWithKey(_cursor.surface, _cursor.dstRect.left, _cursor.dstRect.top,
+			frontBufferScreenSurface, _cursor.srcRect, _cursor.keycolor);
+
+		_cursor.positionChanged = _cursor.surfaceChanged = false;
+		_oldCursorRect = _cursor.dstRect;
+
+		updated = true;
+	}
+
+	return updated;
+}
+
+void AtariGraphicsManager::copySurface8ToSurface16(
+		const Graphics::Surface &srcSurface, const byte *srcPalette,
+		Graphics::Surface &dstSurface, int destX, int destY,
+		const Common::Rect subRect, ScaleMode scaleMode) const {
+	assert(srcSurface.format.bytesPerPixel == 1);
+	assert(dstSurface.format.bytesPerPixel == 2);
+
+	// faster (no memory (re-)allocation) version of Surface::convertTo()
+	const int w = subRect.width();
+	const int h = subRect.height();
+	const Graphics::PixelFormat dstFormat = dstSurface.format;
+
+	const int srcScale = scaleMode == ScaleMode::Downscale ? 2 : 1;
+	const byte *srcRow = (const byte*)srcSurface.getBasePtr(subRect.left * srcScale, subRect.top * srcScale);
+	uint16 *dstRow = (uint16*)dstSurface.getBasePtr(destX, destY);	// already upscaled if needed
+
+	// optimized paths for each case...
+	if (scaleMode == ScaleMode::None) {
+		for (int y = 0; y < h; y++) {
+			for (int x = 0; x < w; x++) {
+				const byte index = *srcRow++;
+
+				*dstRow++ = dstFormat.RGBToColor(
```

Comment:
You might get some performance improvements by using `Graphics::convertPaletteToMap()` to reduce the number of calls to `RGBToColor`.

Reply:
Thanks!